### DISCUSSION TITLE
Opening array from file without knowing its dimensions

### DISCUSSION THREAD

**Discussion by leoyala** [Blosc/c-blosc2#630] [Q&A]
Hello,

I am currently using [b2nd_open](https://www.blosc.org/c-blosc2/reference/b2nd.html?highlight=b2nd_open#c.b2nd_open) to open arrays that I created, for which I know its dimensions. 
But I recently received some arrays for which I do not know its dimensions, and I was wondering if it is possible to load the NDArrays somehow without knowing its dimensions beforehand. Am I correct that NDArrays do not store any information regarding its shape?

**Comment by FrancescAlted** [comment]
I don't see why you should know the dimensions prior to open an array.  Can you send a small program showing what you are trying to do?  (and yes, shape, chunk shape, block shape and the dtype info is stored as metadata; see e.g. "Data types are in!" section in https://www.blosc.org/posts/blosc2-ndim-intro/).

**Reply by leoyala** [reply]
You are right that I should not need the shape of the data to load the `NDArrays`, I was confusing how the opening mechanism works. 
However, I am in the situation where I am saving images to a file that is initialized as [this](https://github.com/IMSY-DKFZ/xilens/blob/develop/src/util.cpp#L22-L56). While I am doing so, I would like to open a second `NDArray` (only read) using `b2nd_open`. Would it be a problem to call `blosc2_init` only once at the beginning of my program and then use `blosc` in multiple threads? One thread is responsible for saving the data, while a different one is responsible for reading other files. In the documentation I see that it is mentioned that in multithreaded environments one can use the [blosc2_compress_ctx](https://www.blosc.org/c-blosc2/reference/context.html#blosc2_8h_1a54d23e475f9039429d6b10848158202b), [blosc2_decompress_ctx](https://www.blosc.org/c-blosc2/reference/context.html#blosc2_8h_1a142c78c63120abd25aae823884ddbbab) pair, but I am not sure if calling `blosc2_init` only once would work instead.

**Reply by FrancescAlted** [reply]
As long as you are careful and use contexts (e.g. `b2nd_create_ctx()`) I think you are safe.  But the best is always try, and in case something wrong happens, then start to scratch your head (and not before).

**Comment by leoyala** [comment]
After trying it out for some time, it seems to work to only initialize blosc in the main thread and then use it from other threads. I have not thoroughly test it, but will close this for now as it seems to work :).
